services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java

```
@@ -0,0 +1,267 @@
+
+        AuthorizationContext reducedContext = new AuthorizationContext()
+                .setAuthorizationRequest(authorizationContext.getAuthorizationRequest())
+                .setTransactionId(authorizationContext.getTransactionId());
+
+        return CorsService.forWebOrigins(authSession)
+                .add(Response.ok(reducedContext));
+    }
+
+    /**
+     * Dereferences request URIs into signed request objects.
+     */
+    @GET
+    @Path(REQUEST_JWT_PATH + "/{requestId}")
+    @Produces(MediaType.TEXT_PLAIN)
+    public Response getSignedRequestObject(@PathParam("requestId") String requestId) {
+        logger.debug("Resolving request URI to signed request object...");
+        AuthorizationContext authorizationContext;
+
+        try {
+            var authSession = this.recoverAuthenticationSession(requestId);
+            authorizationContext = new AuthenticationSessionStore(authSession)
+                    .getAuthorizationContextByRequestId(requestId);
+        } catch (IllegalArgumentException e) {
+            throw new NotFoundException(errorResponse(
+                    Response.Status.NOT_FOUND,
+                    "Authorization context not found for request ID: " + requestId
+            ), e);
+        }
+
+        String requestObjectJwt = authorizationContext.getRequestObjectJwt();
+        return CorsService.open().add(Response.ok(requestObjectJwt));
+    }
+
+    /**
+     * Processes authentication responses from the wallet toward user authentication.
+     */
+    @POST
+    @Path(RESPONSE_URI_PATH)
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response processAuthorizationResponse(
+            @FormParam(ResponseObject.VP_TOKEN_KEY) String vpToken,
+            @FormParam(ResponseObject.PRESENTATION_SUBMISSION_KEY) String presentationSubmission,
+            @FormParam(ResponseObject.STATE_KEY) String state
+    ) {
```

You might validate that vpToken, presentationSubmission, and state are not null or blank before constructing ResponseObject. It would help fail fast with a clear error message if inputs are missing.

---

These are now validated as we construct the response object. Thank you.